### DISCUSSION TITLE
Oddities when running ghostty on Ubuntu

### DISCUSSION THREAD

**Discussion by mikemadden42** [ghostty-org/ghostty#3632] [Q&A]
I *really* love the 1.0.0 release of ghostty.  I'm running in on my main Linux desktop.

```
$ lsb_release -sir
Ubuntu
22.04

$ dpkg -l libgtk-4-dev libadwaita-1-dev git
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                   Version                   Architecture Description
+++-======================-=========================-============-===================================================
ii  git                    1:2.34.1-1ubuntu1.11      amd64        fast, scalable, distributed revision control system
ii  libadwaita-1-dev:amd64 1.1.7-0ubuntu0.22.04.1    amd64        Development files for libadwaita
ii  libgtk-4-dev:amd64     4.6.9+ds-0ubuntu0.22.04.2 amd64        development files for the GTK library
```

```
$ ghostty +version
Ghostty 1.0.0

Version
  - version: 1.0.0
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.6.9
    runtime    : 4.6.9
  - libadwaita : enabled
    build      : 1.1.7
    runtime    : 1.1.7
 ```

Does anyone else see messages like this when running ghostty?

I opened a new tab and get these messages.

```
info(grid): loaded OpenGL 4.6
info(io_exec): found Ghostty resources dir: /home/mjm/.local/share/ghostty
info(io_exec): shell integration automatically injected shell=termio.shell_integration.Shell.zsh
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
warning(gtk): unimplemented action=apprt.action.Action.Key.size_limit
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
info(io_exec): started subcommand path=/bin/sh pid=368912
info(io_exec): subcommand cgroup=-
error(gtk): unable to get current color scheme: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such method “ReadOne”
info(stream): OSC 1 (change icon) received and ignored icon=~
warning(stream): unimplemented OSC command: change_window_icon
info(grid): reallocating GPU buffer old=0 new=20
```

I run the ls command and get these messages.

```
info(grid): reallocating GPU buffer old=20 new=38
info(stream): OSC 1 (change icon) received and ignored icon=ls
warning(stream): unimplemented OSC command: change_window_icon
info(stream): OSC 1 (change icon) received and ignored icon=~
warning(stream): unimplemented OSC command: change_window_icon
warning(stream): unimplemented OSC callback: terminal.osc.Command{ .end_of_command = terminal.osc.Command.Command__struct_90166{ .exit_code = 0 } }
warning(stream): unimplemented OSC command: end_of_command
info(grid): reallocating GPU buffer old=38 new=390
```

I started vim and get these messages.

```
info(stream): OSC 1 (change icon) received and ignored icon=vim
warning(stream): unimplemented OSC command: change_window_icon
warning(stream): ignoring unimplemented CSI 22/23 t
warning(stream): ignoring CSI 22/23 t with extra parameters: ESC [  { 22, 1 } t
info(grid): reallocating GPU buffer old=390 new=6911
info(terminal_dcs): unknown DCS hook: terminal.Parser.Action.DCS{ .intermediates = {  }, .params = {  }, .final = 122 }
warning(stream): unknown CSI m with intermediate: 37
```

**Comment by mikemadden42** [comment]
I also noticed the icon in the dock is a generic icon instead of the ghostty icon.

![Screenshot from 2024-12-27 20-08-06](https://github.com/user-attachments/assets/8655802a-5bda-4322-9649-dd5971353fba)

Since there was not a package for Ubuntu, I [built](https://ghostty.org/docs/install/build) it from source with `zig build -p $HOME/.local -Doptimize=ReleaseFast`.

**Reply by rhodes-b** [reply]
Your GTK and libadwaita dependencies are pretty old (this is expected because you are on ubuntu). Ive only seen people use 4.14 and 1.5 respectfully and even then those versions can have some issues afaik. With 4.16 and 1.6 being the preferred versions can you do

These oddities are probably expected because of those old GTK and adwaita versions

**Reply by sebmestrallet** [reply]
For the dock icon, you can create a [desktop entry file](https://specifications.freedesktop.org/desktop-entry-spec/latest/index.html#introduction) `~/.local/share/applications/ghostty.desktop`:
```
#!/usr/bin/env xdg-open
[Desktop Entry]
Type=Application
Name=Ghostty
Icon=/path/to/ghostty/images/icons/icon_128.png
Exec=/path/to/ghostty/zig-out/bin/ghostty
Terminal=false
StartupNotify=false
Categories=Utility;
```

**Reply by rhodes-b** [reply]
Ghostty already ships with one btw 
https://github.com/ghostty-org/ghostty/blob/6cbd69da7839260508466f9dfb2bc0c0fbb43991/build.zig#L613

**Reply by mikemadden42** [reply]
This dock icon does get setup.  It doesn't seem to work on Ubuntu 22.04, but it works fine on Ubuntu 24.04.

```$ cat .local/share/applications/com.mitchellh.ghostty.desktop
[Desktop Entry]
Name=Ghostty
Type=Application
Comment=A terminal emulator
Exec=ghostty
Icon=com.mitchellh.ghostty
Categories=System;TerminalEmulator;
Keywords=terminal;tty;pty;
StartupNotify=true
Terminal=false
Actions=new-window;
X-GNOME-UsesNotifications=true

[Desktop Action new-window]
Name=New Window
Exec=ghostty
```

**Reply by saiteki-kai** [reply]
I solved this by updating the icon cache:

```bash 
sudo update-icon-caches /usr/share/icons/*
```

**Reply by litemotiv** [reply]
> Your GTK and libadwaita dependencies are pretty old (this is expected because you are on ubuntu). Ive only seen people use 4.14 and 1.5 respectfully and even then those versions can have some issues afaik. With 4.16 and 1.6 being the preferred versions can you do

I'm seeing similar messages and i'm on Fedora with GTK  4.17.4  and libadwaita 1.7. So it doesn't appear to be related to the library versions. E.g. this is when ctrl-tabbing a bit between two Ghostty tabs:

```
feb 14 20:18:48 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:18:50 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:18:50 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:19:03 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:19:03 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:20:33 fedora chronyd[1244]: Selected source 185.45.112.121 (2.fedora.pool.ntp.org)
feb 14 20:21:22 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:21:29 fedora com.mitchellh.ghostty.desktop[26506]: info(io_handler): BELL
feb 14 20:22:11 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:23:00 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:23:01 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:23:04 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:23:04 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:24:09 fedora com.mitchellh.ghostty.desktop[26506]: warning(gtk): unimplemented action=apprt.action.Action.Key.key_sequence
feb 14 20:24:33 fedora com.mitchellh.ghostty.desktop[26506]: warning(stream): unimplemented OSC callback: terminal.osc.Command{ .end_of_command = terminal.osc.Command.Command__struct_97718{ .exit_code = 0 } }
feb 14 20:24:33 fedora com.mitchellh.ghostty.desktop[26506]: warning(stream): unimplemented OSC command: end_of_command
```

**Reply by BrendanMesters** [reply]
I'm having similar errors on Fedora 41 with KDE Plasma 6.3.3

Furthermrore I have kernel 6.12.8 and am running on an Intel NVIDIA setup.

**Reply by pluiedev** [reply]
Please don't revive old threads. You might have a different problem even though it seems like they're related, and if that's the case you just caused ~10 unrelated parties to receive an unwanted email.

**Comment by mikemadden42** [comment]
The visual experience is better on Ubuntu 24.04.1.

```
$ lsb_release -sir
No LSB modules are available.
Ubuntu
24.04

$ ghostty +version
Ghostty 1.0.0

Version
  - version: 1.0.0
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.14.2
    runtime    : 4.14.2
  - libadwaita : enabled
    build      : 1.5.0
    runtime    : 1.5.0
```

![Screenshot from 2024-12-28 07-29-15](https://github.com/user-attachments/assets/da49b4ab-a204-4079-b775-eb929d2c4e4e)

**Comment by mikemadden42** [comment]
Are there plans to support LTS releases of Ubuntu?  Ubuntu 22.04 still has 2+ years of support from Canonical while Ubuntu 24.04 has 4+ years of support.

I would be really nice to support these last couple Ubuntu LTS releases (22.04, 24.04).

**Reply by rhodes-b** [reply]
Its my understanding it does work in 24.04 currently. As for 22.02 I believe there is work being done on a flatpak / you could always try to download / compile more up to date versions of GTK / libadwaita yourself

**Reply by mikemadden42** [reply]
Thanks.  ghostty does work well on Ubuntu 24.04.  It's pretty functional on Ubuntu 22.04 too, outside of missing the custom icon and some extra messages in the console.

**Reply by rhodes-b** [reply]
One of the pepole with more GTK experience would know more than me, my assumption is old packages though

**Reply by mikemadden42** [reply]
The note [here](https://ghostty.org/docs/install/build) makes it sounds like ghostty should work with Ubuntu 23.10 and newer.  It would nice to clarify which versions of Ubuntu are supported since it's a very widely used distribution.

> A recent GTK is required for Ghostty to work with Nvidia (GL) drivers under x11. Ubuntu 22.04 LTS has GTK 4.6 which is not new enough. Ubuntu 23.10 has GTK 4.12 and works. From this discussion the problem was fixed in GTK by Dec 2022. Also, if you are a BTRFS user, make sure to manually upgrade your Kernel (6.6.6 will work). The stock kernel in Ubuntu 23.10 is 6.5.0 which has a bug which causes zig to fail its hash check for packages.

**Reply by CharlieBytesX** [reply]
Sorry for asking here, I can't find how to solve this. 
Im using Pops OS which uses ubuntu 22.04 and x11, I have nvidia... 
How can i upgrade GTK to 4.12 on ubuntu 22.04? I have been searching for days...

Thanks for the help

**Reply by rhodes-b** [reply]
> Sorry for asking here, I can't find how to solve this. 
> Im using Pops OS which uses ubuntu 22.04 and x11, I have nvidia... 
> How can i upgrade GTK to 4.12 on ubuntu 22.04? I have been searching for days...
> 
> Thanks for the help

Compile it yourself is my best answer

**Reply by elerch** [reply]
The appimage build works fine on PopOS. I use it that way without issue: https://github.com/psadi/ghostty-appimage

**Reply by CharlieBytesX** [reply]
Thanks a lot for the help!! Now Im rocking ghostty

**Comment by mikemadden42** [comment]
I experimented with different boxes running Ubuntu 22.04 and ghostty.
- My original desktop was running Ubuntu 22.04 and ghostty.  ghostty was working fine, but I couldn't get the dock icon working.
- I tried ghostty on another two laptops running Ubuntu 22.04.  The dock icon worked fine.
- I tried ghostty on another laptop running Ubuntu 24.04.  The dock icon worked fine.

On the two Ubuntu 22.04 laptops, I was running `x11` for the nvidia driver.  On my original desktop, I only have the built-in video card, and I was running `wayland`.  If I switched to x11 on the desktop, the dock icon for ghostty worked fine.

On the Ubuntu 24.04 laptop, it was running the nvidia driver and wayland.  The dock icon worked fine here.

It seems that for Ubuntu 22.04, x11 + ghostty may be a better setup while wayland + ghostty works fine on Ubuntu 24.04.

I wonder if anyone else comes to this same conclusion for getting ghostty working well on both Ubuntu 22.04 & 24.04?
